zanata-war/src/main/java/org/zanata/service/impl/TranslationServiceImpl.java

```diff
@@ -681,8 +681,8 @@ private String validateTranslations(ContentState newState,
         }
 
         // NB: removedTargets only applies for MergeType.IMPORT
-        final Collection<HTextFlowTarget> removedTargets =
-                new HashSet<HTextFlowTarget>();
+        final Collection<Long> removedTextFlowTargetIds =
+                new HashSet<>();
```

Comment:
Yes, from a memory point of view, this wouldn't work nicely with batches either (since it starts out with a complete set of HTextFlowTargets in the collection). In fact, the separate transactions we added for batches might be the reason equality wasn't working.